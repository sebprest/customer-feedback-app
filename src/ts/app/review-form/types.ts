import { Review } from "^/common/types";
import { ADD_REVIEW } from "./actions";

export type AddReviewAction = {
  type: typeof ADD_REVIEW;
  payload: Review;
};

export type FormState = {
  name: string;
  email: string;
  rating: number;
  comment: string;
};
