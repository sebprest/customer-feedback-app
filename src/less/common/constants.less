@primary: #1e3888;
@white: #ffffff;
@grey: #e5e5e5;

@sm: 768px;
@md: 992px;
@lg: 1200px;

@margin-sm: 8px;
@margin-lg: @margin-sm * 2;

@padding-sm: 8px;
@padding-lg: @padding-sm * 2;

@button-width: 100px;
@button-height: 25px;

@border-radius-sm: 4px;

@font-family: arial, helvetica, sans-serif;
