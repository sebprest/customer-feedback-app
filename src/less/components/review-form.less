.review-form {
  padding: @padding-lg;

  label,
  input {
    display: inline-block;
  }

  label {
    margin-top: @margin-sm;
    margin-left: @margin-sm;
  }

  input {
    display: table;
    width: 80%;
    margin: @margin-sm;
    border-radius: @border-radius-sm;
    border: 1px solid @grey;
  }

  .button {
    width: @button-width;
    height: @button-height;
    color: @white;
    background-color: @primary;
    cursor: pointer;
  }
}
