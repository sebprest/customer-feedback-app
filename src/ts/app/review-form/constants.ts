export const MINIMUM_REVIEW_RATING = 1;
export const MAXIMUM_REVIEW_RATING = 5;

export const REVIEW_FORM_FIELD_IDS = {
  name: "review-form-name",
  email: "review-form-email",
  rating: "review-form-rating",
  comment: "review-form-comment",
};
