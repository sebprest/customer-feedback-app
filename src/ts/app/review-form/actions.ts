import { Review } from "^/common/types";
import { AddReviewAction } from "./types";

export const ADD_REVIEW = "ADD_REVIEW";
export function addReview(newReview: Review): AddReviewAction {
  return {
    type: ADD_REVIEW,
    payload: newReview,
  };
}
