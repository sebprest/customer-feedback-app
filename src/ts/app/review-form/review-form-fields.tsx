import React, { useState } from "react";

import { Review } from "^/common/types";
import {
  MAXIMUM_REVIEW_RATING,
  MINIMUM_REVIEW_RATING,
  REVIEW_FORM_FIELD_IDS,
} from "./constants";

type Props = {
  addReview: (newReview: Review) => void;
};

export const ReviewFormFields: React.FunctionComponent<Props> = React.memo(
  (props) => {
    const defaultValues = { name: "", email: "", rating: 0, comment: "" };
    const [values, setValues] = useState(defaultValues);

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const targetName = event.target.name;
      const targetValue = event.target.value;

      setValues((values) => ({ ...values, [targetName]: targetValue }));
    };

    const onAddReview = (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      props.addReview(values);
      setValues(() => defaultValues);
    };

    return (
      <form className="review-form" onSubmit={onAddReview}>
        <label htmlFor={REVIEW_FORM_FIELD_IDS.name}>Name</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.name}
          type="text"
          name="name"
          value={values.name}
          onChange={handleInputChange}
          required
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.email}>Email</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.email}
          type="text"
          name="email"
          value={values.email}
          onChange={handleInputChange}
          required
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.rating}>Rating</label>
        <input
          type="number"
          id={REVIEW_FORM_FIELD_IDS.rating}
          name="rating"
          value={values.rating}
          onChange={handleInputChange}
          required
          min={MINIMUM_REVIEW_RATING}
          max={MAXIMUM_REVIEW_RATING}
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.comment}>Comment</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.comment}
          name="comment"
          value={values.comment}
          onChange={handleInputChange}
          required
        />
        <input className="button" type="submit" value="Submit" />
      </form>
    );
  }
);

export default ReviewFormFields;
