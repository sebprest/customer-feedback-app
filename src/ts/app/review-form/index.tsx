import React, { useState } from "react";
import { connect } from "react-redux";

import { addReview } from "./actions";
import ReviewFormFields from "./review-form-fields";

type Props = {
  addReview: typeof addReview;
};

export const ReviewForm: React.FunctionComponent<Props> = React.memo(
  (props) => {
    const defaultValues = { name: "", email: "", rating: 0, comment: "" };
    const [values, setValues] = useState(defaultValues);

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const targetName = event.target.name;
      const targetValue = event.target.value;

      setValues((values) => ({ ...values, [targetName]: targetValue }));
    };

    const onAddReview = (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      props.addReview(values);
      setValues(() => defaultValues);
    };

    return (
      <ReviewFormFields
        addReview={onAddReview}
        updateFormState={handleInputChange}
        formState={values}
      />
    );
  }
);

export default connect(null, { addReview })(ReviewForm);
