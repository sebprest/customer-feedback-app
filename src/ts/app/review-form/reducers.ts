import { Review, ReviewsState } from "^/common/types";
import { ADD_REVIEW } from "./actions";
import { AddReviewAction } from "./types";

export function review(
  state: Review[] = [],
  action: AddReviewAction
): ReviewsState {
  switch (action.type) {
    case ADD_REVIEW:
      return [...state, action.payload];
    default:
      return state;
  }
}
