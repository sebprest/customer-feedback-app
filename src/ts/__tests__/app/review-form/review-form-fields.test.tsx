import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import { REVIEW_FORM_FIELD_IDS } from "^/app/review-form/constants";

import { ReviewFormFields } from "^/app/review-form/review-form-fields";
import { FAKE_REVIEW } from "^/__tests__/__helpers__/fake-review";

describe("ReviewFormFields", () => {
  const commonProps = {
    addReview: jest.fn(),
    updateFormState: jest.fn(),
    formState: FAKE_REVIEW,
  };
  it("should render the fields for the review form", () => {
    const reviewFormFields = renderer.create(
      <ReviewFormFields {...commonProps} />
    );
    expect(reviewFormFields).toMatchSnapshot();
  });

  it("should call the addReview action when the form is submitted", () => {
    const addReview = jest.fn().mockReturnValue(Promise.resolve());
    const reviewFormFields = mount(
      <ReviewFormFields {...commonProps} addReview={addReview} />
    );
    reviewFormFields.find(".review-form").simulate("submit");
    expect(addReview).toHaveBeenCalledTimes(1);
  });

  it("should call the updateFormState action when a field is updated", () => {
    const updateFormState = jest.fn().mockReturnValue(Promise.resolve());
    const reviewFormFields = mount(
      <ReviewFormFields {...commonProps} updateFormState={updateFormState} />
    );
    reviewFormFields.find(`#${REVIEW_FORM_FIELD_IDS.name}`).simulate("change");
    expect(updateFormState).toHaveBeenCalledTimes(1);
  });
});
