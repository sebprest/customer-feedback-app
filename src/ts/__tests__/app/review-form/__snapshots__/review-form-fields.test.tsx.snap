// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewFormFields should render the fields for the review form 1`] = `
<form
  className="review-form"
  onSubmit={[Function]}
>
  <label
    htmlFor="review-form-name"
  >
    Name
  </label>
  <input
    id="review-form-name"
    name="name"
    onChange={[Function]}
    required={true}
    type="text"
    value=""
  />
  <label
    htmlFor="review-form-email"
  >
    Email
  </label>
  <input
    id="review-form-email"
    name="email"
    onChange={[Function]}
    required={true}
    type="text"
    value=""
  />
  <label
    htmlFor="review-form-rating"
  >
    Rating
  </label>
  <input
    id="review-form-rating"
    name="rating"
    onChange={[Function]}
    value={0}
    required={true}
  />
  <label
    htmlFor="review-form-comment"
  >
    Comment
  </label>
  <input
    id="review-form-comment"
    name="comment"
    onChange={[Function]}
    required={true}
    value=""
  />
  <input
    type="submit"
    value="Submit"
  />
</form>
`;
